File: 2_Data_Structures/week3_hash_tables/3_hash_substring/HashSubstring.java
import java.io.BufferedOutputStream;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

public class HashSubstring {

    private static FastScanner in;
    private static PrintWriter out;
    private static int PRIME = 1000000007;
    private static int MULTIPLIER = 263;

    private static long hashFunc(String s) {
        long hash = 0;
        for (int i = s.length() - 1; i >= 0; --i)
            hash = (hash * MULTIPLIER + s.charAt(i)) % PRIME;
        return hash;
    }

    private static long[] precomputeHashes(String text, String pattern) {
        int tLen = text.length();
        int pLen = pattern.length();
        long[] hVal = new long[tLen - pLen + 1];
        long y = 1;
        for (int i = 0; i < pLen; ++i)
            y = (y * MULTIPLIER) % PRIME;
        hVal[hVal.length - 1] = hashFunc(text.substring(tLen - pLen, tLen));
        for (int i = tLen - pLen - 1; i >= 0; --i) {
            hVal[i] = (MULTIPLIER * hVal[i + 1] + text.charAt(i) - y * text.charAt(i + pLen)) % PRIME;

            // Fix negative modulo value
            if (hVal[i] < 0)
                hVal[i] += PRIME;
        }
        return hVal;
    }

    public static void main(String[] args) throws IOException {
        in = new FastScanner();
        out = new PrintWriter(new BufferedOutputStream(System.out));
        printOccurrences(getOccurrences(readInput()));
        out.close();
    }

    private static Data readInput() throws IOException {
        String pattern = in.next();
        String text = in.next();
        return new Data(pattern, text);
    }

    private static void printOccurrences(List<Integer> ans) throws IOException {
        for (Integer cur : ans) {
            out.print(cur);
            out.print(" ");
        }
    }

//    private static List<Integer> getOccurrences(Data input) {
//        String s = input.pattern, t = input.text;
//        int m = s.length(), n = t.length();
//        List<Integer> occurrences = new ArrayList<Integer>();
//        for (int i = 0; i + m <= n; ++i) {
//            boolean equal = true;
//            for (int j = 0; j < m; ++j) {
//                if (s.charAt(j) != t.charAt(i + j)) {
//                    equal = false;
//                    break;
//                }
//            }
//            if (equal)
//                occurrences.add(i);
//        }
//        return occurrences;
//    }

    private static List<Integer> getOccurrences(Data input) {
        String pattern = input.pattern, text = input.text;
        int pLen = pattern.length(), tLen = text.length();
        List<Integer> occurrences = new ArrayList<Integer>();
        long pHash = hashFunc(pattern);
        long hVal[] = precomputeHashes(text, pattern);
        for (int i = 0; i < (tLen - pLen + 1); ++i) {
            if (pHash == hVal[i]) {
                // Check to see if the strings are really equal
                if (pattern.equals(text.substring(i, i + pLen)))
                    occurrences.add(i);
            }
        }
        return occurrences;
    }

    static class Data {
        String pattern;
        String text;

        public Data(String pattern, String text) {
            this.pattern = pattern;
            this.text = text;
        }
    }

    static class FastScanner {
        private BufferedReader reader;
        private StringTokenizer tokenizer;

        public FastScanner() {
            reader = new BufferedReader(new InputStreamReader(System.in));
            tokenizer = null;
        }

        public String next() throws IOException {
            while (tokenizer == null || !tokenizer.hasMoreTokens()) {
                tokenizer = new StringTokenizer(reader.readLine());
            }
            return tokenizer.nextToken();
        }

        public int nextInt() throws IOException {
            return Integer.parseInt(next());
        }
    }
}

File: 2_Data_Structures/week3_hash_tables/3_hash_substring/hash_substring.py
# python3
DEBUG = False
DEBUG_HASH = False

def read_input():
    return (input().rstrip(), input().rstrip())

def print_occurrences(output):
    print(' '.join(map(str, output)))

# implement string hashing with polynomial hash function
# and Rabin-Karp's algorithm to find all occurrences of
# the pattern in the text
# define a large prime
_prime = 1000000007
# define a multiplier
_multiplier = 263
# define a hash function
def poly_hash(s):
    ans = 0
    for c in reversed(s):
        ans = (ans * _multiplier + ord(c)) % _prime
    return ans

# precompute the hashes of all substrings of the text
def precompute_hashes(text, pLen):
    # initialize an empty list of hashes
    hashes = [None] * (len(text) - pLen + 1)
    # compute the hash of the last |pattern| characters of the text
    hashes[len(text) - pLen] = poly_hash(text[len(text) - pLen:])
    # precompute the value of multiplier^(|pattern|) % prime
    factor = 1
    for _ in range(pLen):
        factor = (factor * _multiplier) % _prime
    # iterate through the text, computing the hash of each |pattern| characters
    for i in range(len(text) - pLen - 1, -1, -1):
        # update the hash of the |pattern| characters of the text
        # but don't use pow() because it's slow
        hashes[i] = (_multiplier * hashes[i + 1] + ord(text[i]) - factor * ord(text[i + pLen])) % _prime
        # check hash[i] against poly_hash(text[i:i + pLen])
        if DEBUG_HASH:
            assert(hashes[i] == poly_hash(text[i:i + pLen]))
    # return hashes
    return hashes

def get_occurrences(pattern, text):
    """
    Find all occurrences of a pattern in a text using the Rabin-Karp algorithm.

    Args:
        pattern (str): The pattern to search for.
        text (str): The text to search in.

    Returns:
        list: A list of indices where the pattern occurs in the text.
    
    Examples:
        >>> get_occurrences('aba', 'abacaba')
        [0, 4]
        >>> get_occurrences('Test', 'testTesttesT')
        [4]
        >>> get_occurrences('aaaaa', 'baaaaaaa')
        [1, 2, 3]    
    """
    # initialize an empty list of occurrences
    occurrences = []
    # compute the hash of the pattern
    p_hash = poly_hash(pattern)
    # get the hashes of all substrings of the text
    hashes = precompute_hashes(text, len(pattern))
    # compute the hash of the first |pattern| characters of the text
    pLen = len(pattern)
    tLen = len(text)
    # iterate through the text, checking for a match
    for i in range(tLen - pLen + 1):
        # check if the hashes match
        if p_hash != hashes[i]:
            continue
        # check if the strings match
        if text[i:i + pLen] == pattern:
            # append the index to occurrences
            occurrences.append(i)
    # return occurrences
    return occurrences

# write test_cases
def test_cases():
    assert get_occurrences('aba', 'abacaba') == [0, 4]
    assert get_occurrences('Test', 'testTesttesT') == [4]
    assert get_occurrences('aaaaa', 'baaaaaaa') == [1, 2, 3]
    print('all test cases passed...')

if __name__ == '__main__':
    if DEBUG:
        test_cases()
    else:
        print_occurrences(get_occurrences(*read_input()))

